lib/search_files.js

```diff
@@ -0,0 +1,252 @@
+const fs = require('fs');
+const path = require('path');
+const oshomedir = require('os-homedir');
+
+const isFunction = require('lodash.isfunction');
+const isPlainObject = require('lodash.isplainobject');
+const isArray = require('lodash.isarray');
+const isString = require('lodash.isstring');
+const doNothing = require('lodash.identity');
+const mapValues = require('lodash.mapvalues');
+
+const isAbsolutePath = require('is-absolute');
+const parsePath = require('parse-filepath');
+
+const isWindows = process.platform === 'win32';
+
+
+function firstUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var ret = fn.apply(arr, [arr[i]].concat(args));
+    if (ret) {
+      return ret;
+    }
+  }
+  return null;
+}
+
+function eachUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = [];
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var v = fn.apply(arr, [arr[i]].concat(args));
+    if (v) {
+      if (isArray(v)) {
+        Array.prototype.push.apply(ret, v);
+      } else {
+        ret.push(v);
+      }
+    }
+  }
+  return ret;
+}
+
+
+function normalizeHomedir(pth, homedir) {
+  if (!pth) {
+    return pth;
+  }
+
+  if (pth[0] === '~') {
+    if (pth.length === 1) {
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
+      return path.join(homedir, pth.slice(2));
+    }
+
+    if (isWindows && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
+    }
+  }
+
+  return pth;
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
+    pth = pth();
+  }
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
+    return path.join(ctx.cwd, pth);
+  }
+
+  return path.normalize(pth);
+}
+
+
+var initCtx = {
+  homedir: oshomedir(),
+  cwd: process.cwd(),
+};
+
+function normalizeContext(ctx) {
+  ctx = mapValues(ctx || {}, doNothing);
+
+  ctx.cwd = !ctx.cwd ? initCtx.cwd :
+    normalizePath(ctx.cwd, initCtx);
+
+  ctx.homedir = !ctx.homedir ? initCtx.homedir :
+    normalizePath(ctx.homedir, initCtx);
+
+  return ctx;
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
```

Comment:
I didn't think these files triggered an auto require of the loaders.  If that is the case, there's a lot more work that needs to go into this and I think it would increasingly slow it down.